src/en/charms-deploying.md

```diff
@@ -281,43 +281,63 @@ are supplied, then the remaining units will be assigned as normal to a new,
-your space first before deploying a charm that attempts to use it.
+More complex networks can be configured using spaces.  Spaces group one or more
+routable subnets with common ingress and egress rules to give the operator much
+better and finer-grained control over all networking aspects of a model and its
+application deployments.
 
-When `--bind` is not used, all endpoints will use the host machine's preferred
-private address, which you can see using `unit-get private-address`.
+See the [How to configure more complex networks using spaces][spaces] for
+details on creating and listing spaces.
 
-To deploy mysql while binding the "db" endpoint to an address that is part
-of the "database" space:
+When deploying a charm or a bundle, you can specify a space using the `--bind`
+argument following the `juju deploy` command.
+
+When deploying an application to a target with multiple spaces, the operator
+must specify which space to use - ambiguous bindings will result in a
+provisioning failure. The following, for example, will deploy the 'mysql'
+application to the 'db-space' space:
 
 ```bash
-juju deploy mysql --bind db=database
+juju deploy mysql --bind db-space
 ```
 
-To deploy wordpress while binding all endpoints of wordpress to the
-"internal-apps" space:
+For finer control, the `--bind` argument can also be used to specify how
+specific charm-defined endpoints are connected to specific spaces, including a
+default option for any interfaces not specified:
 
 ```bash
-juju deploy wordpress --bind internal-apps
+juju deploy --bind "db:db-space db-admin:admin-space default-space" mysql
 ```
 
-To deploy haproxy while binding "url" to "public" and all other endpoints to
-"internal":
+For information on building bundles with bindings, see [Using and Creating
+Bundles][creatingbundles].
+
+Both the `add-machine` and `deploy` commands allow the specification of a
+spaces constraint using the `--constraints` argument:
 
 ```bash
-juju deploy haproxy --bind "url=public internal"
+juju add-machine --constraints spaces=db-space
 ```
 
-To use binding with bundles, see the related section in
-[Using and creating bundles](./charms-bundles.html).
+The spaces constraint allows you to select an instance for the new machine or
+unit, connected to one or more existing spaces. Both positive and negative
+entries are accepted, the latter prefixed by "^", in a comma-delimited list.
+For example, given the following:
 
-To learn about `extra-bindings`, which provide a way to declare an extra
-bindable endpoint that is not a relation, see [Charm metadata](./authors-charm-metadata.html).
+```
+--constraints spaces=db-space,^storage,^dmz,internal
+```
+
+Juju will provision instances connected to (with IP addresses on) one of the
+subnets of both db-space and internal spaces, and NOT connected to either the storage
+or dmz spaces.
+
+For more information regarding constraints in general, see "juju help
```

Comment:
juju help topics are deprecated - but there is a reference page on constraints: ./reference-constraints